src/txHashBuilder.c

```diff
@@ -572,142 +673,150 @@ void txHashBuilder_addPoolRegistrationCertificate_enterRelays(tx_hash_builder_t*
-        const uint16_t* port,
-        const ipv4_t* ipv4,
-        const ipv6_t* ipv6
-)
+static void _relay_addPort(tx_hash_builder_t* builder, ipport_t* port)
 {
-	_TRACE("state = %u, remainingRelays = %u", builder->state, builder->poolCertificateData.remainingRelays);
+	_TRACE("state = %d, remainingRelays = %u", builder->state, builder->poolCertificateData.remainingRelays);
 
 	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS);
-	ASSERT(builder->poolCertificateData.remainingRelays > 0);
-	builder->poolCertificateData.remainingRelays--;
 
-	// Array(4)[
-	//   Unsigned[0]
 	//   Unsigned[port] / Null
+	if (port->isNull) {
+		BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+	} else {
+		BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, port->number);
+	}
+}
+
+static void _relay_addIpv4(tx_hash_builder_t* builder, ipv4_t* ipv4)
+{
+	_TRACE("state = %d, remainingRelays = %u", builder->state, builder->poolCertificateData.remainingRelays);
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS);
+
 	//   Bytes[ipv4] / Null
+	if (ipv4->isNull) {
+		BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+	} else {
+		STATIC_ASSERT(sizeof(ipv4->ip) == IPV4_SIZE, "wrong ipv4 size"); // SIZEOF does not work for 4-byte buffers
+		BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, IPV4_SIZE);
+		BUILDER_APPEND_DATA(ipv4->ip, IPV4_SIZE);
+	}
+}
+
+static void _relay_addIpv6(tx_hash_builder_t* builder, const ipv6_t* ipv6)
+{
+	_TRACE("state = %d, remainingRelays = %u", builder->state, builder->poolCertificateData.remainingRelays);
+
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS);
+
 	//   Bytes[ipv6] / Null
-	// ]
-	{
-		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 4);
-		{
-			BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, 0);
-		}
-		{
-			if (port != NULL) {
-				BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, *port);
-			} else {
-				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
-			}
-		}
-		{
-			if (ipv4 != NULL) {
-				STATIC_ASSERT(sizeof(ipv4->ip) == IPV4_SIZE, "wrong ipv4 size"); // SIZEOF does not work for 4-byte buffers
-				BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, IPV4_SIZE);
-				BUILDER_APPEND_DATA(ipv4->ip, IPV4_SIZE);
-			} else {
-				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
-			}
-		}
-		{
-			if (ipv6 != NULL) {
-				STATIC_ASSERT(SIZEOF(ipv6->ip) == IPV6_SIZE, "wrong ipv6 size");
-				BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, IPV6_SIZE);
-
-				// serialized as 4 big-endian uint32
-				STATIC_ASSERT(SIZEOF(ipv6->ip) == 16, "wrong ipv6 size");
-				uint32_t* as_uint32 = (uint32_t*) ipv6->ip;
-				for (size_t i = 0; i < 4; i++) {
-					uint8_t chunk[4];
-					u4be_write(chunk, as_uint32[i]);
-					BUILDER_APPEND_DATA(chunk, 4);
-				}
-			} else {
-				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
-			}
+	if (ipv6->isNull) {
+		BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+	} else {
+		STATIC_ASSERT(SIZEOF(ipv6->ip) == IPV6_SIZE, "wrong ipv6 size");
+		BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, IPV6_SIZE);
+
+		// serialized as 4 big-endian uint32
```

Comment:
why was that needed? wouldn't it be more straightforward to treat the ipv6 addresses as a byte array of 16 bytes all the time?